inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sinks/filecollect/SenderManager.java

```diff
@@ -0,0 +1,454 @@
+
+        proxyClientConfig.setIoThreadNum(ioThreadNum);
+        proxyClientConfig.setEnableBusyWait(enableBusyWait);
+        proxyClientConfig.setProtocolType(ProtocolType.TCP);
+
+        SHARED_FACTORY = new DefaultThreadFactory("agent-sender-manager-" + sourcePath,
+                Thread.currentThread().isDaemon());
+
+        DefaultMessageSender sender = new DefaultMessageSender(proxyClientConfig, SHARED_FACTORY);
+        sender.setMsgtype(msgType);
+        sender.setCompress(isCompress);
+        this.sender = sender;
+    }
+
+    public void sendBatch(SenderMessage message) {
+        while (!resendQueue.isEmpty()) {
+            AgentUtils.silenceSleepInMs(retrySleepTime);
+        }
+        addAckInfo(message.getAckInfo());
+        sendBatchWithRetryCount(message, 0);
+    }
+
+    private void addAckInfo(PackageAckInfo info) {
+        packageAckInfoLock.writeLock().lock();
+        packageAckInfoList.add(info);
+        packageAckInfoLock.writeLock().unlock();
+    }
+
+    public boolean sendFinished() {
+        boolean finished = false;
+        packageAckInfoLock.writeLock().lock();
+        if (packageAckInfoList.isEmpty()) {
+            finished = true;
+        }
+        packageAckInfoLock.writeLock().unlock();
+        return finished;
+    }
+
+    /**
+     * flushOffset
+     */
+    private void doFlushOffset() {
+        packageAckInfoLock.writeLock().lock();
+        PackageAckInfo info = null;
+        for (int i = 0; i < packageAckInfoList.size();) {
+            if (packageAckInfoList.get(i).getHasAck()) {
+                info = packageAckInfoList.remove(i);
+                MemoryManager.getInstance().release(AGENT_GLOBAL_WRITER_PERMIT, info.getLen());
+            } else {
+                break;
+            }
+        }
+        if (info != null) {
+            LOGGER.info("save offset {} taskId {} instanceId {}", info.getOffset(), profile.getTaskId(),
+                    profile.getInstanceId());
+            OffsetProfile offsetProfile = new OffsetProfile(profile.getTaskId(), profile.getInstanceId(),
+                    info.getOffset(), profile.get(INODE_INFO));
+            offsetManager.setOffset(offsetProfile);
+        }
+        packageAckInfoLock.writeLock().unlock();
+    }
+
+    private void clearOffset() {
+        packageAckInfoLock.writeLock().lock();
+        for (int i = 0; i < packageAckInfoList.size();) {
+            MemoryManager.getInstance().release(AGENT_GLOBAL_WRITER_PERMIT, packageAckInfoList.remove(i).getLen());
+        }
+        packageAckInfoLock.writeLock().unlock();
+    }
+
+    /**
+     * Send message to proxy by batch, use message cache.
+     */
+    private void sendBatchWithRetryCount(SenderMessage message, int retry) {
+        boolean suc = false;
+        while (!suc) {
+            try {
+                // suc = true;
+                // AgentSenderCallback cb = new AgentSenderCallback(message, retry);
```

Comment:
It is recommended to remove the commented code

Reply:
get